lom/src/engine/engine.go

```diff
@@ -0,0 +1,207 @@
+                    break
+                case <- p.chClReadAbort:
+                    return
+                }
+            } else {
+                LogError("Failed to read request. err(%v)", err)
+                return
+            }
+        }
+    }()
+}
+
+
+func (p *engine_t) runLoop() {
+    /*
+     * Wait for
+     *      signal to refresh
+     *      request from client
+     *      Internal timer for outstanding request's/seq's timeout processing
+     */
+
+    /* To abort periodic log at the end of loop */
+    chAbortLog := make(chan interface{}, 1)
+    LogPeriodicInit(chAbortLog)
+
+    defer func() {
+        /* Indicate loop end */
+        p.chTrack <- 1
+
+        /* Abort LogPeriodic */
+        chAbortLog <- 0
+
+        if len(p.chClReadAbort) < cap(p.chClReadAbort) {
+            p.chClReadAbort <- "Aborted"
+        }
+
+        close(p.chTrack)
+        close(chAbortLog)
+        close(p.chClReadAbort)
+    }()
+
+    /* Handle signal for config update & terminate */
+    chSignal := make(chan os.Signal, 3)
+    signal.Notify(chSignal, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM)
+
+    /*
+     * kick off reading client requests
+     *
+     * p.chClientReq - The read loop sends read requests via this chan to main loop.
+     * p.chClReadAbort - Way to abort the read loop.
+     */
+    p.chClientReq = make(chan *LoMRequestInt, 1)
+    p.chClReadAbort = make(chan interface{}, 2)
+    p.readRequest()
+
+    /*
+     * Initialize seq handler
+     *
+     * The seq handler is invoked for any response from clients via this 
+     * runLoop Go routine. To ensure that any timer based processing too
+     * happens in the same context, it uses this routine for any of its
+     * timer call back needs.
+     * On any async timer firing, it intimates this loop via chSeqHandler channel
+     * and the main loop invokes sequence handler.
+     *
+     * This way any logical processing by sequence handler is via the context
+     * of single Go routine only.
+     */
+    chSeqHandler := make(chan int64, 2)
+    InitSeqHandler(chSeqHandler)
+
+    server := GetServerReqHandler()
+
+    /* Intimate the start of loop */
+    p.chTrack <- 0
+loop:
+    for {
+        select {
+        case msg := <- p.chClientReq:
+            server.processRequest(msg)
+
+        case <- chSeqHandler:
+            GetSeqHandler().processTimeout()
+
+        case sig := <- chSignal:
+            switch(sig) {
+            case syscall.SIGHUP, syscall.SIGINT:
+                /*
+                 * Reload.
+                 * NOTE: Any currently active sequence will not be affected
+                 * On any error, continues to use last loaded values. 
+                 */
+                 InitConfigPath(cfgPath)
+
+            case syscall.SIGTERM:
+                break loop
+            }
+        case <- p.chTerminate:
+            break loop
+        }
+    }
+} 
+
+func (p *engine_t) close() {
```

Comment:
Is synchronization needed here while len() & cap() calls are executed ?

Reply:
No, as this is the only writer.